streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamsPartitionAssignor.java

```diff
@@ -646,11 +677,21 @@ public void onAssignment(final Assignment assignment) {
                 partitionsByHost = info.partitionsByHost();
                 break;
             case 3:
+                final int latestSupportedVersionGroupLeader = info.latestSupportedVersion();
+                if (latestSupportedVersionGroupLeader > usedSubscriptionMetadataVersion) {
+                    final int newSubscriptionMetadataVersion = Math.min(latestSupportedVersionGroupLeader, SubscriptionInfo.LATEST_SUPPORTED_VERSION);
+                    log.info("Sent a version {} subscription and group leader's latest supported version is {}. " +
+                        "Upgrading subscription metadata version to {} for next rebalance.",
+                        usedSubscriptionMetadataVersion,
+                        latestSupportedVersionGroupLeader,
+                        newSubscriptionMetadataVersion);
+                    usedSubscriptionMetadataVersion = newSubscriptionMetadataVersion;
```

Comment:
Should we enforce another rebalance when this happens? Otherwise we may live in a long period where all clients can support a higher version now while we still use lower version, so some advanced features (e.g. rack aware assignment in the future) cannot be picked, until the next rebalance is triggered.

Reply:
Maybe -- my original ideas was, to recommend users to do "N + 1" rolling bounce (or do the last bounce in a delayed fashion: ie, stop last instance to be upgraded -- wait for rebalance to finish -- restart the instance. The last step will not result in version probing, as all other instances are "ready" for the new version already and the new version would be used.

Let me know what you think.

Reply:
If we can handle it automatically be code, I'd suggest we avoid enforcing users to do that manually. Do you have any concerns of always doing that in code, is there any corner cases this may not be desirable?

Reply:
There is one corner case: if you have `N` instances, and `N-1` got upgraded, if the `N`th instances fails, the first `N-1` would upgrade themselves to the new protocol. On restart/recovery of the `N`th instance, it would join as "old" instance resulting in a downgrade for all others. (This downgrade scenario could happen for other circumstances, too, and should not be an issue though.)

I am not against doing it in code. Just not sure if the "N + 1" bounce strategy is a big burden (would like to avoid unnecessary code complexity). If we do it in code, I am wondering if only the leader should behave like this?

Also note, that this case would only happen, if the leader is bounced last (if I didn't miss something). For all other cases, after upgrading the leader the following rolling bounces would already trigger this upgrade. Thus, in practice, the N+1 rebalance might only be necessary in rare cases.

Reply:
Note that we decrement `usedSubscriptionMetadataVersion` when received leader assignment indicating version probing, and only increment `usedSubscriptionMetadataVersion` here after the leader assignment indicate that all client has at least some version larger, i.e. `latestSupportedVersionGroupLeader > usedSubscriptionMetadataVersion`. So my understanding is that, no matter if the leader is bounced last or not, after we bounce each instance exactly once, i.e. N times in total, then only in the N-ith rebalance the leader will recognize the new version, AND realize that everyone can now support the new version, and hence return the assignment with the incremented `latestSupportedVersionGroupLeader`.

Putting it in other words, if we bounce each instance once, then in each of the N-1 bounces, the version probing will actually causes 2 rebalances as the bounced instance will always get that it needs to decrement its version and trigger another rebalance, while the last bounce will only trigger 1 rebalance, as described above. So we are effectively triggering `2N - 1` rebalances, right (we can validate if my reasoning is right from the current system test)? It sounds to me that triggering `2N` rebalances to complete the whole process is a reasonable choice.

I think one special handling we can consider, is that if with the newer version, the new version assignment does not depend on the new version subscription (which is indeed the case for this bump-up), then we can short-cut the last rebalance, by letting the leader to do this: `when I saw everyone can indeed support version 3, although their sent subscription is in version 2, I will force sending the assignment in version 3 knowing that it does not actually rely on any extra information of subscription in version 3`. But admittedly it introduce quite some complexity into the streams partition assignor so maybe not worth it.

Reply:
Thanks for detailed comments. Sounds right. 

About sending a version 3 assignment if all support version 3 (even if they send version 2): for this upgrade, if subscription info does actually not change it would work. However, I am wondering if we should stick with current implementation to strictly decouple different versions. Otherwise, it's getter more complex: In the future, we might not be able---due to missing information in the subscription---to send the new assignment back and thus we would end up in different behavior depending on the source/target version of the upgrade. It might be better to be conservative and assume "incompatibility" all the time to get consistent behavior.

Still wondering, if each instance should trigger the additional N+1 rebalance, or only the leader. Thoughts?

One more question for my own education: when we have `N` instances and the leader is bounced as instance `X`, ie, `1...X-1,X,X+1,...N`, how is the new leader chosen? From my understanding, if will be one of `X+1...N` because all other instances are "younger"? This seems to imply, that the leader is upgrades at last, because new leader will always be chosen from old instances? Is this correct?